docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+application state is the one that corresponds to the blockchain at
+height *terminationHeight*.
+
+#### **[FS-VC-BLOCKS-INV]**:
+If *FastSync* terminates successfully, then the
+returned list of blocks  is the one that corresponds to the blocks of
+the
+blockchain.
+
+
+> As this specification does
+> not assume that a correct peer is at the most recent height
+> of the blockchain (it might lag behind), the property [FS-SEQ-SAFE-START]
+> cannot be ensured in an unreliable distributed setting. We consider
+> the following relaxation. (Which is typically sufficient for
+> Tendermint, as the consensus reactor then synchronizes from that
+> height.)
+
+#### **[FS-VC-CORR-INV]**:
+Under [FS-ALL-CORR-PEER], let *maxh* be the maximum 
+height of a correct peer [**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link]
+in *peerIDs* at the time *Fastsync* starts. If *FastSync* terminates
+successfully, it is at some height *terminationHeight >= maxh - 1*.
+
+> The above property is independent of how many blocks are added to the
+> blockchain (and learned by the peers) while *Fastsync* is running. It
+> links the target height to the initial state. If *Fastsync* has to
+> catch-up many blocks, it would be better to  link the
+> target height to a time close to the termination. This is captured by
+> the following specification:
+
+
+#### **[FS-VC-CORR-INV-SYNC]**:
+Under [FS-ALL-CORR-PEER], there exists a constant time interval *TD*, such
+that if *term* is the time *Fastsync* terminates and
+*maxh* be the maximum height of a correct peer
+[**[TMBC-CORR-FULL]**][TMBC-CORR-FULL-link] in *peerIDs* at the time
+*term - TD*, then if *FastSync* terminates successfully, it is at
+some height *terminationHeight >= maxh*.
+
+> We use *term - TD* as reference time, as we have to account
+> for communication delay between the peer and *FS*. After the peer sent
+> the last message to *FS*, the peer and *FS* run concurrently and
+> independently. There is no assumption on the rate at which a peer can
+> add blocks (e.g., it might be in the process of catching up itself).
+
+>  (*TD* might depend on timeouts etc. We suggest that an acceptable
+>  value for *TD* is in the range of apporx. 10 sec., that is the
+>  intervall between two calls `QueryStatus()`; see below.
+
+
+> Under [FS-ALL-CORR-PEER], if *peerIDs* contains a full node that is
+> "synchronized with the blockchain", and *blockchainheight* is the height
+> of the blockchain at time *term*, then  *terminationHeight* may even
+> achieve
+> *blockchainheight - TD / ETIME*;
+> cf. [**[TMBC-SEQ-APPEND-E]**][TMBC-SEQ-APPEND-E-link]. 
+
+
+
+
+
+#### Liveness
+
+<!--
+> liveness specifications in English. Possibly with timing/fairness requirements:
+e.g., if the component is connected to a correct full node and communication is
+reliable and timely, then something good happens eventually. 
+---->
+
+#### **[FS-VC-ALL-CORR-TERM]**:
+Under [FS-ALL-CORR-PEER], *Fastsync* eventually terminates successfully.
```

Comment:
Where do we define what we mean by termination with failure?

Reply:
In the following. Is this OK?

#### **[FS-V2-TIMEOUT]**:

*Fastsync V2* starts a timeout whenever a block is
executed (that is, when the height is incremented). If the timeout expires
before the next block is executed, *Fastsync* terminates.
We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates
with failure, otherwise it terminates successfully.